gemini/sample-apps/llamaindex-rag/src/app/main.py

```diff
@@ -0,0 +1,484 @@
+import json
+import logging
+import os
+from pathlib import Path
+from typing import Any, Iterator, List, Optional, Sequence, Tuple, Union, cast
+import uuid
+
+from common.common import DATA_PATH
+from common.utils import download_blob
+from datasets import Dataset
+from fastapi import Depends, FastAPI, HTTPException
+from google.api_core.client_options import ClientOptions
+from google.api_core.exceptions import PermissionDenied
+import google.auth
+import google.auth.transport.requests
+from google.cloud import (
+    aiplatform,
+    firestore,
+    firestore_admin_v1,
+    secretmanager,
+    storage,
+)
+from google.oauth2 import service_account
+from langchain_google_vertexai import ChatVertexAI, VertexAIEmbeddings
+import pandas as pd
+from pydantic import BaseModel
+
+# Ragas must be 0.9.1
+from ragas import evaluate
+from ragas.metrics import (
+    answer_correctness,
+    answer_relevancy,
+    answer_similarity,
+    context_precision,
+    context_recall,
+    context_relevancy,
+    faithfulness,
+)
+from src.rag.evaluate import LLMEvaluator, write_results_to_bq
+import uvicorn
+import yaml
+
+logging.basicConfig(
+    filename="eval.log", encoding="utf-8", level=logging.INFO
+)  # Set the desired logging level
+logger = logging.getLogger(__name__)
+
+config_path = os.path.join(
+    os.path.dirname(__file__), "..", "..", "common", "config.yaml"
+)
+with open(config_path, "r") as config_file:
+    config = yaml.safe_load(config_file)
+
+# Now you can use the processed paths
+PROJECT_ID = config["project_id"]
+SERVICE_ACCOUNT_KEY = config["service_account_key"]
+FIRESTORE_DB_NAME = config.get("firestore_db_name")
+
+
+ragas_metrics_dict = {
+    "context_precision": context_precision,
+    "answer_relevancy": answer_relevancy,
+    "faithfulness": faithfulness,
+    "context_relevancy": context_relevancy,
+    "context_recall": context_recall,
+    "answer_similarity": answer_similarity,
+    "answer_correctness": answer_correctness,
+}
+
+
+def get_secret(project_id, secret_id, version_id="latest"):
+    client = secretmanager.SecretManagerServiceClient()
+    name = f"projects/{project_id}/secrets/{secret_id}/versions/{version_id}"
+    response = client.access_secret_version(name=name)
+    return response.payload.data.decode("UTF-8")
+
+
+class IndexUpdate(BaseModel):
+    base_index_name: str
+    base_endpoint_name: str
+    qa_index_name: Optional[str]
+    qa_endpoint_name: Optional[str]
+    firestore_db_name: Optional[str]
+    firestore_namespace: Optional[str]
+
+
+class PromptUpdate(BaseModel):
+    prompt_name: str
+    new_content: str
+
+
+class RAGConfig(BaseModel):
+    llm_name: str = "gemini-1.5-flash"
+    temperature: float = 0.2
+    similarity_top_k: int = 5
+    retrieval_strategy: str = "auto_merging"
+    use_hyde: bool = True
+    use_refine: bool = True
+    use_node_rerank: bool = True
+    use_react: bool = True
+    qa_followup: bool = True
+    hybrid_retrieval: bool = True
+
```

Comment:
Do you need all of these custom classes? Shouldn't there already be some in the libraries you're using?

Reply:
Yes we need these to represent requests to FastAPI backend. These make it much easier to maintain contracts between frontend and backend.